x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts

```diff
@@ -83,3 +105,31 @@ export const PrebuiltRuleAsset = BaseCreateProps.omit(BASE_PROPS_REMOVED_FROM_PR
       version: RuleVersion,
     })
   );
+
+/**
+ * Creates a Map of the fields that are upgradable during the Upgrade workflow, by type.
+ * Creating this Map dynamically, based on BaseCreateProps and TypeSpecificFields, ensures that we don't need to:
+ *  - manually add rule types to this Map if they are created
+ *  - manually add or remove any fields if they are added or removed to a specific rule type
+ *  - manually add or remove any fields if we decide that they should not be part of the upgradable fields.
+ */
+function createUpgradableRuleFieldsByTypeMap() {
+  const baseFields = Object.keys(
+    BaseCreateProps.omit(BASE_PROPS_REMOVED_FROM_PREBUILT_RULE_ASSET).shape
+  );
+
+  const specificTypesToOmit: readonly string[] = TYPE_SPECIFIC_FIELDS_TO_OMIT;
+
+  return new Map(
+    TypeSpecificCreateProps.options.map((option) => {
+      const typeName = option.shape.type.value;
+      const typeSpecificFields = Object.keys(option.shape).filter(
+        // Filter out type-specific fields that should not be part of the upgradable fields
+        (field) => !specificTypesToOmit.includes(field)
+      );
+      return [typeName, [...baseFields, ...typeSpecificFields]];
+    })
+  );
+}
+
+export const UPGRADABLE_RULES_FIELDS_BY_TYPE_MAP = createUpgradableRuleFieldsByTypeMap();
```

Comment:
Not sure I understand how and where this map is gonna be used, can you elaborate?

Also, what I'm missing in this PR is changes to the `upgrade/_perform` request schema. We talked that its `fields` property's schema should be derived from the rule schema, I think this needs to be done in this PR:

https://github.com/elastic/kibana/blob/71b90a1af4df7ea2b3fb51244b4cf3ffec70d7a9/x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/perform_rule_upgrade/perform_rule_upgrade_route.ts#L89-L133

Reply:
> Also, what I'm missing in this PR is changes to the upgrade/_perform request schema. We talked that its fields property's schema should be derived from the rule schema, I think this needs to be done in this PR:

Yes, you're right, I'm trying to do some Zod acrobatics to get this schema created. It's WIP.